exonum-java-binding/testkit/src/test/java/com/exonum/binding/testkit/TestKitTest.java

```diff
@@ -31,115 +33,222 @@
-    List<Class<? extends ServiceModule>> serviceModules = ImmutableList.of(TestServiceModule.class,
-        TestServiceModule.class);
     TestKit.Builder testKitBuilder = TestKit.builder()
-        .withServices(serviceModules);
-    assertThrows(exceptionType, testKitBuilder::build);
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        .withArtifactsDirectory(artifactsDirectory);
+    IllegalArgumentException thrownException = assertThrows(exceptionType, testKitBuilder::build);
+    assertThat(thrownException.getMessage())
+        .isEqualTo("Following service artifacts were deployed, but not used for"
+            + " service instantiation: [%s]", ARTIFACT_ID.toString());
+  }
+
+  @Test
+  void createTestKitWithoutDeployedArtifactThrows() {
+    Class<IllegalArgumentException> exceptionType = IllegalArgumentException.class;
+    TestKit.Builder testKitBuilder = TestKit.builder()
+        .withArtifactsDirectory(artifactsDirectory);
+    IllegalArgumentException thrownException = assertThrows(exceptionType,
+        () -> testKitBuilder.withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID));
+    assertThat(thrownException.getMessage())
+        .isEqualTo("Service %s should be deployed first in order to be created",
+        ARTIFACT_ID.toString());
   }
 
   @Test
-  void createTestKitWithBuilderForMultipleDifferentServices() {
+  void createTestKitWithDifferentDeployedArtifactAndCreatedServiceThrows() {
+    Class<IllegalArgumentException> exceptionType = IllegalArgumentException.class;
+    TestKit.Builder testKitBuilder = TestKit.builder()
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        .withArtifactsDirectory(artifactsDirectory);
+    IllegalArgumentException thrownException = assertThrows(exceptionType,
+        () -> testKitBuilder.withService(ARTIFACT_ID_2, SERVICE_NAME_2, SERVICE_ID_2));
+    assertThat(thrownException.getMessage())
+        .isEqualTo("Service %s should be deployed first in order to be created",
+        ARTIFACT_ID_2.toString());
+  }
+
+  @Test
+  void createTestKitWithoutArtifactsDirectoryThrows() {
+    Class<NullPointerException> exceptionType = NullPointerException.class;
+    TestKit.Builder testKitBuilder = TestKit.builder()
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        .withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID);
+    NullPointerException thrownException = assertThrows(exceptionType, testKitBuilder::build);
+    assertThat(thrownException.getMessage()).isEqualTo("Artifacts directory was not set.");
+  }
+
+  @Test
+  void createTestKitWithBuilderForSingleServiceWithCustomConfiguration() {
     try (TestKit testKit = TestKit.builder()
-        .withService(TestServiceModule.class)
-        .withService(TestServiceModule2.class)
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        .withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID, DEFAULT_CONFIGURATION)
```

Comment:
This test actually passes the same `Any.defaultInstance` as the testkit by default, to the service that *ignores* the configuration. I.e., if we change the line/implementation this test won't fail.

Reply:
Created a task for that - https://jira.bf.local/browse/ECR-3652